File: PoolWall.java
import java.awt.Color;
public class PoolWall extends Wall
{
   private boolean collidingSide; // if true and wall is verticle ball collides from left, if true and wall is horizontal ball collides from bottom

   public PoolWall(double x, double y, double l, boolean isV, Color c, boolean cs)
   {
      super(x, y, l, isV, c);
      collidingSide = cs;
   }
   @Override
   public boolean checkCollisions(Entity b)
   {
      if(b instanceof Ball)
      {
         if(this.isVert())
         {
            if(collidingSide)//ball hits wall from left
            {
               if(b.getX() + ((Ball)b).getRadius() >= this.getX() && b.getY() >= this.getY() && b.getY() <= this.getY() + this.length())
                  return true;
            
            }
            else //ball hits wall from right
            {
               if(b.getX() - ((Ball)b).getRadius() <= this.getX() && b.getY() >= this.getY() && b.getY() <= this.getY() + this.length())
                  return true;
            
            }
            if((b.getX()-this.getX()) * (b.getX()-this.getX()) + (b.getY() - (this.getY() + this.length())) * (b.getY() - (this.getY()+this.length())) < ((Ball)b).getRadius() * ((Ball)b).getRadius())  //Ball hits top edge
               return true;
         
         }
         else
         {
            if(collidingSide)//hits from bottom
            {
               if(b.getY() + ((Ball)b).getRadius() >= this.getY() && b.getX() >= this.getX() && b.getX() <= this.getX() + this.length())
                  return true;
            }
            else //hits from top
            {
               if(b.getY() - ((Ball)b).getRadius() <= this.getY() && b.getX() >= this.getX() && b.getX() <= this.getX() + this.length())
                  return true;
            }
            if((b.getX()-(this.getX()+this.length())) * (b.getX()-(this.getX()+this.length())) + (b.getY() - this.getY()) * (b.getY() - this.getY()) < ((Ball)b).getRadius() * ((Ball)b).getRadius())
               return true;
         }
         if((b.getX()-this.getX()) * (b.getX()-this.getX()) + (b.getY() - this.getY()) * (b.getY() - this.getY()) < ((Ball)b).getRadius() * ((Ball)b).getRadius())
            return true;
      
      
      }
      else
         return false;
      return false;
   }
   
   public void impact(Entity b)
   {
   if(b instanceof Ball)
      {
         if(this.isVert())
         {
            if(collidingSide)//ball hits wall from left
            {
               if(b.getX() + ((Ball)b).getRadius() >= this.getX() && b.getY() >= this.getY() && b.getY() <= this.getY() + this.length())
               {
               double dx = ((Ball)b).getRadius() - (this.getX() - b.getX());
               double dt = dx/(b.getXV());
               double dy = b.getYV() * dt;
                  
                  b.setX(b.getX() - dx); 
                  b.setY(b.getY() - dy);
                  b.setXV(b.getXV() * -1);
               }
               
            
            }
            else //ball hits wall from right
            {
               if(b.getX() - ((Ball)b).getRadius() <= this.getX() && b.getY() >= this.getY() && b.getY() <= this.getY() + this.length())
                  {
                  double dx = ((Ball)b).getRadius() - (b.getX() - this.getX());
                  double dt = dx/(b.getXV());
                  double dy = b.getYV() * dt;
                  
                  b.setX(b.getX() + dx); 
                  b.setY(b.getY() - dy);
                  b.setXV(b.getXV() * -1);                  
                  }
            
            }
            if((b.getX()-this.getX()) * (b.getX()-this.getX()) + (b.getY() - (this.getY() + this.length())) * (b.getY() - (this.getY()+this.length())) < ((Ball)b).getRadius() * ((Ball)b).getRadius())  //Ball hits top edge
               {
               ((Ball)b).impactPoint(this.getX(), this.getY() + this.length());
               }
         
         }
         else
         {
            if(collidingSide)//hits from bottom
            {
               if(b.getY() + ((Ball)b).getRadius() >= this.getY() && b.getX() >= this.getX() && b.getX() <= this.getX() + this.length())
                  {
                  double dy = ((Ball)b).getRadius() - (this.getY() - b.getY());
                  double dt = dy/(b.getYV());
                  double dx = b.getXV() * dt;
                  b.setX(b.getX() - dx);
                  b.setY(b.getY() - dy);
                  b.setYV(b.getYV() * -1);
                  }
            }
            else //hits from top
            {
               if(b.getY() - ((Ball)b).getRadius() <= this.getY() && b.getX() >= this.getX() && b.getX() <= this.getX() + this.length())
                  {
                  double dy = ((Ball)b).getRadius() - (b.getY() - this.getY());
                  double dt = dy/(b.getYV());
                  double dx = b.getXV() * dt;
                  b.setX(b.getX() - dx);
                  b.setY(b.getY() + dy);
                  b.setYV(b.getYV() * -1);
                  }
            }
            if((b.getX()-(this.getX()+this.length())) * (b.getX()-(this.getX()+this.length())) + (b.getY() - this.getY()) * (b.getY() - this.getY()) < ((Ball)b).getRadius() * ((Ball)b).getRadius())
               {
               ((Ball)b).impactPoint(this.getX() + this.length(), this.getY());
               }
         }
         if((b.getX()-this.getX()) * (b.getX()-this.getX()) + (b.getY() - this.getY()) * (b.getY() - this.getY()) < ((Ball)b).getRadius() * ((Ball)b).getRadius())
         {
         ((Ball)b).impactPoint(this.getX(), this.getY());
         }
      
      
      }
   
   
   }

}

File: PoolWall.kt
import java.awt.Color

class PoolWall(
    x: Double,
    y: Double,
    l: Double,
    isV: Boolean,
    c: Color?, // if true and wall is vertical ball collides from left, if true and wall is horizontal ball collides from bottom
    private val collidingSide: Boolean
) : Wall(x, y, l, isV, c) {
    override fun checkCollisions(b: Entity?): Boolean {
        if (b is Ball) {
            if (this.isVert) {
                if (collidingSide) //ball hits wall from left
                {
                    if (b.x + b.radius >= x && b.y >= y && b.y <= y + length()) return true
                } else  //ball hits wall from right
                {
                    if (b.x - b.radius <= x && b.y >= y && b.y <= y + length()) return true
                }
                if ((b.x - x) * (b.x - x) + (b.y - (y + length())) * (b.y - (y + length())) < b.radius * b.radius) //Ball hits top edge
                    return true
            } else {
                if (collidingSide) //hits from bottom
                {
                    if (b.y + b.radius >= y && b.x >= x && b.x <= x + length()) return true
                } else  //hits from top
                {
                    if (b.y - b.radius <= y && b.x >= x && b.x <= x + length()) return true
                }
                if ((b.x - (x + length())) * (b.x - (x + length())) + (b.y - y) * (b.y - y) < b.radius * b.radius) return true
            }
            if ((b.x - x) * (b.x - x) + (b.y - y) * (b.y - y) < b.radius * b.radius) return true
        } else return false
        return false
    }

    override fun impact(b: Entity?) {
        if (b is Ball) {
            if (this.isVert) {
                if (collidingSide) //ball hits wall from left
                {
                    if (b.x + b.radius >= x && b.y >= y && b.y <= y + length()) {
                        val dx = b.radius - (x - b.x)
                        val dt = dx / b.xV
                        val dy = b.yV * dt
                        b.x = b.x - dx
                        b.y = b.y - dy
                        b.xV = b.xV * -1
                    }
                } else  //ball hits wall from right
                {
                    if (b.x - b.radius <= x && b.y >= y && b.y <= y + length()) {
                        val dx = b.radius - (b.x - x)
                        val dt = dx / b.xV
                        val dy = b.yV * dt
                        b.x = b.x + dx
                        b.y = b.y - dy
                        b.xV = b.xV * -1
                    }
                }
                if ((b.x - x) * (b.x - x) + (b.y - (y + length())) * (b.y - (y + length())) < b.radius * b.radius) //Ball hits top edge
                {
                    b.impactPoint(x, y + length())
                }
            } else {
                if (collidingSide) //hits from bottom
                {
                    if (b.y + b.radius >= y && b.x >= x && b.x <= x + length()) {
                        val dy = b.radius - (y - b.y)
                        val dt = dy / b.yV
                        val dx = b.xV * dt
                        b.x = b.x - dx
                        b.y = b.y - dy
                        b.yV = b.yV * -1
                    }
                } else  //hits from top
                {
                    if (b.y - b.radius <= y && b.x >= x && b.x <= x + length()) {
                        val dy = b.radius - (b.y - y)
                        val dt = dy / b.yV
                        val dx = b.xV * dt
                        b.x = b.x - dx
                        b.y = b.y + dy
                        b.yV = b.yV * -1
                    }
                }
                if ((b.x - (x + length())) * (b.x - (x + length())) + (b.y - y) * (b.y - y) < b.radius * b.radius) {
                    b.impactPoint(x + length(), y)
                }
            }
            if ((b.x - x) * (b.x - x) + (b.y - y) * (b.y - y) < b.radius * b.radius) {
                b.impactPoint(x, y)
            }
        }
    }
}